pkg/services/ngalert/migration/ualert.go

```diff
@@ -4,89 +4,93 @@ import (
+	legacymodels "github.com/grafana/grafana/pkg/services/alerting/models"
 	migmodels "github.com/grafana/grafana/pkg/services/ngalert/migration/models"
-	migrationStore "github.com/grafana/grafana/pkg/services/ngalert/migration/store"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 )
 
-func (om *OrgMigration) migrateAlerts(ctx context.Context, alerts []*migrationStore.DashAlert, info migmodels.DashboardUpgradeInfo) ([]*AlertPair, error) {
+func (om *OrgMigration) migrateAlerts(ctx context.Context, alerts []*legacymodels.Alert, info migmodels.DashboardUpgradeInfo) ([]models.AlertRule, error) {
 	log := om.log.New(
 		"dashboardUid", info.DashboardUID,
 		"dashboardName", info.DashboardName,
 		"newFolderUid", info.NewFolderUID,
 		"newFolderNane", info.NewFolderName,
 	)
 
-	pairs := make([]*AlertPair, 0, len(alerts))
+	rules := make([]models.AlertRule, 0, len(alerts))
 	for _, da := range alerts {
 		al := log.New("ruleID", da.ID, "ruleName", da.Name)
 		alertRule, err := om.migrateAlert(ctx, al, da, info)
 		if err != nil {
 			return nil, fmt.Errorf("migrate alert: %w", err)
 		}
-		pairs = append(pairs, &AlertPair{AlertRule: alertRule, DashAlert: da})
+		rules = append(rules, *alertRule)
 	}
 
-	return pairs, nil
+	return rules, nil
 }
 
-func (om *OrgMigration) migrateDashboard(ctx context.Context, dashID int64, alerts []*migrationStore.DashAlert) ([]*AlertPair, error) {
+func (om *OrgMigration) migrateDashboard(ctx context.Context, dashID int64, alerts []*legacymodels.Alert) ([]models.AlertRule, error) {
 	info, err := om.migratedFolder(ctx, om.log, dashID)
 	if err != nil {
 		return nil, fmt.Errorf("get or create migrated folder: %w", err)
 	}
-	pairs, err := om.migrateAlerts(ctx, alerts, *info)
+	rules, err := om.migrateAlerts(ctx, alerts, *info)
 	if err != nil {
 		return nil, fmt.Errorf("migrate and save alerts: %w", err)
 	}
 
-	return pairs, nil
+	return rules, nil
 }
 
-func (om *OrgMigration) migrateOrgAlerts(ctx context.Context) ([]*AlertPair, error) {
+func (om *OrgMigration) migrateOrgAlerts(ctx context.Context) error {
 	mappedAlerts, cnt, err := om.migrationStore.GetOrgDashboardAlerts(ctx, om.orgID)
 	if err != nil {
-		return nil, fmt.Errorf("load alerts: %w", err)
+		return fmt.Errorf("load alerts: %w", err)
 	}
 	om.log.Info("Alerts found to migrate", "alerts", cnt)
 
-	pairs := make([]*AlertPair, 0, cnt)
 	for dashID, alerts := range mappedAlerts {
-		dashPairs, err := om.migrateDashboard(ctx, dashID, alerts)
+		rules, err := om.migrateDashboard(ctx, dashID, alerts)
 		if err != nil {
-			return nil, fmt.Errorf("migrate and save dashboard '%d': %w", dashID, err)
+			return fmt.Errorf("migrate and save dashboard '%d': %w", dashID, err)
+		}
+
+		if len(rules) > 0 {
+			om.log.Debug("Inserting migrated alert rules", "count", len(rules))
+			err := om.migrationStore.InsertAlertRules(ctx, rules...)
+			if err != nil {
+				return fmt.Errorf("insert alert rules: %w", err)
+			}
```

Comment:
insert is moved to where it is supposed to be because now generation of rules and routes are implicitly related via the migration logic, right?

Reply:
Correct, before we couldn't save the alerts until we migrated the channels since they added labels. Now they are independent, so we can insert somewhere more natural.